---
title: "Visualization of Ovarian Cancer DATA using various R packages"
knit: (function(input_file, encoding) {
    out_dir<- 'docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
output: html_document
---

```{r}
library(curatedOvarianData) 
data(package="curatedOvarianData")
```

```{r}
data(TCGA_eset) 
# Get gene names
gene_names <- featureNames(TCGA_eset)
head(gene_names, 10)



```

## Extract metadata and clinical information:
we can access this information using the pData() and fData() functions from the Biobase package.
```{r}
summary(TCGA_eset)
dim(TCGA_eset)
#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("Biobase")

library(Biobase)
sample_metadata <- pData(TCGA_eset)
#head(sample_metadata)
feature_metadata <- fData(TCGA_eset)
#head(feature_metadata)
```

The TCGA_eset dataset contains gene expression data for 13,104 features (genes) across 578 samples. The data has been preprocessed and normalized using a standardized pipeline to remove batch effects and other sources of technical variability. The resulting dataset contains 20,501 genes and is stored as an ExpressionSet object in R, which includes not only the gene expression data but also sample annotation information and other metadata.

# Heatmap using pheatmap

```{r}

library(curatedOvarianData)
library(pheatmap)


# Extract gene expression data
exprs_matrix <- exprs(TCGA_eset)

# Set row names to gene names
rownames(exprs_matrix) <- featureNames(TCGA_eset)

# Subset to top 100 most variable genes by custom filter
exprs_matrix <- exprs_matrix[apply(exprs_matrix, 1, function(x) var(x) > 0.5), ][1:100, ]

# Create heatmap
pheatmap(exprs_matrix, scale = "row", clustering_distance_rows = "euclidean", show_rownames = FALSE)

```

# correlation plot using corplot library

```{r}
library(corrplot)
gene_expression <- exprs(TCGA_eset)
# Select a subset of genes (e.g., the first 100 genes)
selected_genes <- gene_expression[1:100, ]

# Select a subset of samples (e.g., the first 50 samples)
selected_samples <- selected_genes[, 1:50]

correlation_matrix_subset <- cor(t(selected_samples), method = "pearson")

corrplot(correlation_matrix_subset, method = "color", type = "upper", mar = c(0, 0, 1, 1), tl.pos = "n")


```

Network analysis plots: Network analysis can be used to identify the relationships between different metabolites and their role in metabolic pathways. The "igraph" package in R can be used to generate network analysis plots that show the connections between metabolites.

```{r}
library(igraph)
# Load the TCGA_eset dataset
data(TCGA_eset)

# Extract the gene expression data
tcga.exprs <- exprs(TCGA_eset)

# Log2 transformation of the gene expression data
log2.exprs <- log2(tcga.exprs + 1)

# Calculate the variance for each gene
gene.variance <- apply(log2.exprs, 1, var)

# Order genes by variance and select the top 50
top.genes <- head(order(gene.variance, decreasing = TRUE), 50)

# Subset the gene expression data with the top 50 genes
subset.exprs <- log2.exprs[top.genes, ]

# Generate a correlation matrix of the gene expression data
cor.matrix <- cor(subset.exprs)

# Create an igraph graph object
graph <- graph.adjacency(cor.matrix, mode = "undirected", weighted = TRUE)

# Set node color based on gene expression level
node.color <- ifelse(rowMeans(subset.exprs) > log2(6), "red", "blue")

# Plot the graph with node color based on gene expression level
plot(graph, vertex.size =8, vertex.color = node.color, edge.width = E(graph)$weight * 2, vertex.label=NA)



```


## volcano plot using ggplot
Volcano plots are commonly used in omics data analysis, including metabolomics, to visualize the significance and fold-change of features between two groups or conditions. 
These plots are especially useful for identifying features that are significantly different between groups.
```{r}
library(limma)
# Create a design matrix using your sample group information

design <- model.matrix(~ 0 + sample_metadata$sample_type)

# Fit the linear model
fit <- lmFit(TCGA_eset, design)

# Apply empirical Bayes statistics
fit <- eBayes(fit)
summary(fit)

results <- topTable(fit, number = nrow(TCGA_eset), adjust = "fdr", sort.by = "none")
head(results)
```

```{r}
library(ggplot2)
# Prepare the data for the volcano plot
volcano_data <- data.frame(
  genes = results$gene,
  log2FoldChange = results$sample_metadata.sample_typetumor - results$sample_metadata.sample_typehealthy,
  pvalue = results$adj.P.Val
)

# Create the volcano plot
ggplot(volcano_data, aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(aes(color = pvalue < 0.05), alpha = 0.7) +
  scale_color_manual(values = c("blue", "red")) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  labs(x = "Log2 Fold Change", y = "-log10 adjusted p-value", title = "Differentially Expressed Genes") +
  theme_classic()

```

## complex heatmap

```{r}
library(ComplexHeatmap)

top_gene_expression <- exprs(TCGA_eset)
sample_metadata <- pData(TCGA_eset)
top_gene_expression_t <- t(top_gene_expression)
combined_data <- data.frame(Sample_Type = sample_metadata$sample_type, top_gene_expression_t)
#non_na_indices contain the indices of all samples that have a non-missing value
non_na_indices <- which(!is.na(sample_metadata$sample_type))
# we create a new data frame with the Sample_Type column, using only the rows corresponding to the non_na_indices.
sample_annotation_filtered <- data.frame(Sample_Type = sample_metadata$sample_type[non_na_indices])
rownames(sample_annotation_filtered) <- rownames(sample_metadata)[non_na_indices]
# filter the columns of the top_gene_expression matrix to include only those that correspond to the non_na_indices
top_gene_expression_filtered <- top_gene_expression[, non_na_indices]


```

```{r}
col_fun <- colorRampPalette(c("blue", "white", "red"))(100)

Heatmap(top_gene_expression_filtered,
        col = col_fun,
        name = "Expression",
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        show_row_names = FALSE,
        top_annotation = HeatmapAnnotation(df = sample_annotation_filtered$Sample_Type))

```

## circlize 
The plot is constructed using the circlize package in R, which allows for circular visualization of network data. The color of the edges represents the superpathway of the first metabolite (numerator of the ratio) of the metabolite pair that constructs the metabolite ratios. The grayscale gradient filling the edges indicates the strength of the genetic association with darker color indicating stronger significance.
```{r}

```

