---
title: "Visualization using various R packages"
knit: (function(input_file, encoding) {
    out_dir<- 'docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
output: html_document
---

```{r}
suppressWarnings(library(curatedOvarianData)) 
data(package="curatedOvarianData")
```

The TCGA_eset dataset contains gene expression data for 13,104 features (genes) across 578 samples. The data has been preprocessed and normalized using a standardized pipeline to remove batch effects and other sources of technical variability. The resulting dataset is stored as an Expression Set object in R, which includes not only the gene expression data but also sample annotation information and other metadata.
```{r}
data(TCGA_eset) 
# Get gene names
gene_names <- featureNames(TCGA_eset)
head(gene_names, 10)



```



## Extract metadata and clinical information:
we can access this information using the pData() and fData() functions from the Biobase package.

```{r}
#summary(TCGA_eset)
dim(TCGA_eset)

library(Biobase)
sample_metadata <- pData(TCGA_eset)
#head(sample_metadata)
feature_metadata <- fData(TCGA_eset)
#head(feature_metadata)
```



## Heatmaps 

```{r}

suppressWarnings(library(curatedOvarianData))
suppressWarnings(library(pheatmap))


# Extract gene expression data
exprs_matrix <- exprs(TCGA_eset)

rownames(exprs_matrix) <- featureNames(TCGA_eset)

# Subset to top 20 genes that have a relatively high variance
exprs_matrix <- exprs_matrix[apply(exprs_matrix, 1, function(x) var(x) > 0.5), ][20:40, ]

# Create heatmap
pheatmap(exprs_matrix, scale = "row", clustering_distance_rows = "euclidean", show_rownames = T, show_colnames=F)

```
**The plot shows a heatmap of the expression levels of the top 20 most variable genes in a subset of samples. Each row represents a gene, and each column represents a sample. The color scale represents the level of expression for each gene in each sample, with red indicating high expression and blue indicating low expression. The rows are clustered using euclidean distance and the columns are not clustered**



```{r}
suppressWarnings(library(ComplexHeatmap))

top_gene_expression <- exprs(TCGA_eset)
sample_metadata <- pData(TCGA_eset)
top_gene_expression_t <- t(top_gene_expression)
combined_data <- data.frame(Sample_Type = sample_metadata$sample_type, top_gene_expression_t)
#non_na_indices contain the indices of all samples that have a non-missing value
non_na_indices <- which(!is.na(sample_metadata$sample_type))
# we create a new data frame with the Sample_Type column, using only the rows corresponding to the non_na_indices.
sample_annotation_filtered <- data.frame(Sample_Type = sample_metadata$sample_type[non_na_indices])
rownames(sample_annotation_filtered) <- rownames(sample_metadata)[non_na_indices]
# new expression matrix that includes the columns for samples with non-missing Sample_Type values.
top_gene_expression_filtered <- top_gene_expression[, non_na_indices]


```

```{r}
col_fun <- colorRampPalette(c("blue", "white", "red"))(100)

Heatmap(top_gene_expression_filtered,
        col = col_fun,
        name = "Expression",
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        show_row_names = FALSE,
        #add annotations to the heatmap based on the sample type column(healthy and tumor)
        top_annotation = HeatmapAnnotation(data = sample_annotation_filtered$Sample_Type))

```
**The heatmap is showing the gene expression levels for the top variable genes in the dataset, with rows representing genes and columns representing individual samples. The color scale represents the relative expression levels of the genes, with red indicating high expression and blue indicating low expression.  The samples are annotated with their sample type (tumor or healthy).This allows for visual inspection of any potential patterns or differences in gene expression between different sample types.**



## Correlation Plot 

```{r}
suppressWarnings(library(corrplot))
gene_expression <- exprs(TCGA_eset)
# Select a subset of genes (e.g., the first 20 genes)
selected_genes <- gene_expression[1:20, ]

# Select a subset of samples (e.g., the first 100 samples)
selected_samples <- selected_genes[, 1:100]

correlation_matrix_subset <- cor(t(selected_samples), method = "pearson")

corrplot(correlation_matrix_subset, method = "color", type = "upper", mar = c(0, 0, 1, 1))


```
**This plot is showing the correlation between the expression levels of a subset of genes (the first 20 genes) in a subset of samples (the first 100 samples). The plot is a correlation matrix, where each cell represents the correlation between the expression levels of two genes. The color of each cell indicates the strength of the correlation, with darker colors indicating stronger positive or negative correlations.**

```{r}
suppressWarnings(library(corrplot))
gene_expression <- exprs(TCGA_eset)
# Select a subset of genes (e.g., the first 100 genes)
selected_genes <- gene_expression[1:100, ]

# Select a subset of samples (e.g., the first 100 samples)
selected_samples <- selected_genes[, 1:100]

correlation_matrix_subset <- cor(t(selected_samples), method = "pearson")

corrplot(correlation_matrix_subset, method = "color", type = "upper", mar = c(0, 0, 1, 1),tl.pos="n")


```


## Network visualization
```{r}
suppressWarnings(library(igraph))
# Load the TCGA_eset dataset
data(TCGA_eset)

# Extract the gene expression data
tcga.exprs <- exprs(TCGA_eset)
# Subset the gene expression data to the first 30 samples
tcga.exprs <- tcga.exprs[, 1:30]

#log2 transformation can help to normalize the range of expression values and make the graph more informative.
log2.exprs <- log2(tcga.exprs + 1)

# Calculate the variance for each gene
gene.variance <- apply(log2.exprs, 1, var)

# Order genes by variance and select the top 100
top.genes <- head(order(gene.variance, decreasing = TRUE), 100)

# Subset the gene expression data with the top 50 genes
subset.exprs <- log2.exprs[top.genes, ]

# Generate a correlation matrix of the gene expression data
cor.matrix <- cor(subset.exprs)

# Create an igraph graph object
graph <- graph.adjacency(cor.matrix, mode = "undirected", weighted = TRUE)
# Remove self-loops
graph <- simplify(graph, remove.multiple = FALSE, remove.loops = TRUE)
# Set node color based on gene expression level
node.color <- ifelse(rowMeans(subset.exprs) > log2(6), "red", "blue")

# Plot the graph with node color based on gene expression level
plot(graph, vertex.size =8, vertex.color = node.color,vertex.label.cex = 0.7, edge.width = E(graph)$weight*2)



```


**Each node represents a gene.The edges connecting the nodes represent the correlation between the expression levels of the connected genes. The edge width is proportional to the correlation strength.The node color is determined by the average gene expression level. If the mean expression level is greater than log2(6), the node is colored red, otherwise, it's colored blue.**




```{r}

suppressWarnings(library(affy))
suppressWarnings(library(circlize))

data(TCGA_eset)

# Extract gene expression data
gene_expression <- exprs(TCGA_eset)

# Calculate variance for each gene
gene_variance <- apply(gene_expression, 1, var)

# Get the indices of the top 20 genes with the highest variance
top_genes <- order(gene_variance, decreasing = TRUE)[1:20]

# Subset the gene expression data with the top 20 genes
subset_exprs <- gene_expression[top_genes, ]

# Calculate the correlation matrix of the gene expression data
correlation_matrix <- cor(t(subset_exprs))

grid.col <- setNames(rainbow(length(top_genes)), top_genes)


chordDiagram(correlation_matrix, annotationTrack = "grid", 
             preAllocateTracks = 1, 
             grid.col = grid.col,
             directional = 1, 
             direction.type = c("diffHeight", "arrows"), 
             link.arr.type = "big.arrow")

circos.trackPlotRegion(track.index = 1, panel.fun = function(x, y) {
  xlim = get.cell.meta.data("xlim")
  ylim = get.cell.meta.data("ylim")
  sector.name = get.cell.meta.data("sector.index")
  circos.text(CELL_META$xcenter, 
              ylim[1] + cm_h(2), 
              sector.name, 
              facing = "clockwise",
              niceFacing = TRUE, 
              adj = c(0, 0.5),
              cex = .5,
              col=grid.col[sector.name],
              font = 1)
  circos.axis(h = "bottom",
              labels.cex = .1,
              sector.index = sector.name
              )
}, bg.border = NA)
```


**we are using a chord diagram to visualize the gene expression data of the top 6 genes and 6 selected samples. The chord diagram shows the relationship between the genes and samples by connecting them with chords. The thickness of the chords represents the strength of the correlation between the genes and samples.The colors are just used to differentiate between the different gene-sample connections in the plot. Each gene is represented by a unique color, and each connection between a gene and a sample is represented by a segment with that gene's color.** 


```{r}

suppressWarnings(library(GEOquery))
suppressWarnings(library(networkD3))

# Download the GSE113690 dataset
gse <- getGEO("GSE113690")
eset <- gse[[1]]

# Extract gene expression data
exprs <- exprs(eset)

# Calculate variance for each gene
gene_variance <- apply(exprs, 1, var)

# Get the indices of the top 20 genes with the highest variance
top_genes <- head(order(gene_variance, decreasing = TRUE), 20)

# Randomly select 20 samples
set.seed(1)
selected_samples <- sample(ncol(exprs), 20)

# Subset the gene expression data with the top 20 genes and the selected 20 samples
subset_exprs <- exprs[top_genes, selected_samples]

# Calculate correlation matrix
cor_mat <- cor(subset_exprs)

# Convert correlation matrix to edge list
edge_list <- as.data.frame(as.table(cor_mat))
colnames(edge_list) <- c("source", "target", "weight")

# Create the simple network

simpleNetwork(edge_list,linkColour = "blue")
            

```

**This plot is showing a simple network diagram where nodes represent the top 20 genes with the highest variance and edges represent the correlation between them based on the subset of gene expression data. The thickness of the edges represents the strength of the correlation between the genes.**


```{r}
suppressWarnings(library(visNetwork))
library(visNetwork)
# Create a data frame with nodes and edges
nodes <- data.frame(id = 1:10, 
                    label = paste0("Node ", 1:10), 
                    value = sample(1:10, 10, replace = TRUE),
                    color = c("red", "blue", "green", "orange", "purple", 
                              "yellow", "pink", "brown", "grey", "black"))
edges <- data.frame(from = sample(1:10, 20, replace = TRUE), 
                    to = sample(1:10, 20, replace = TRUE))

# Create the network visualization
visNetwork(nodes, edges) %>% 
  visOptions(highlightNearest = TRUE, 
             nodesIdSelection = TRUE) %>% 
  visLayout(randomSeed = 123)


```


```{r}
suppressWarnings(library(nycflights13))
suppressWarnings(library(visNetwork))
suppressWarnings(library(dplyr))

# subset only 10 flights from each origin airport
flights_subset <- flights %>%
  group_by(origin) %>%
  sample_n(size = 10)

# create nodes and edges data frames
nodes <- data.frame(id = unique(c(flights_subset$origin, flights_subset$dest)), 
                    label = unique(c(flights_subset$origin, flights_subset$dest)))
edges <- data.frame(from = flights_subset$origin, to = flights_subset$dest)

# create the network visualization
visNetwork(nodes, edges, width = "100%") %>%
  visNodes(shape = "dot", size = 10) %>%
  visEdges(arrows = "to", smooth = TRUE) %>%
  visLayout(randomSeed = 123) 

```


## Diagram
```{r}
suppressWarnings(library(DiagrammeR))

grViz("
digraph {
  
  # Set node defaults
  node [shape = diamond, style = filled, color = skyblue, fontname = 'Helvetica']

  # Add nodes to the graph
  A [label = 'Gene A']
  B [label = 'Gene B']
  C [label = 'Gene C']
  D [label = 'Gene D']
  E [label = 'Gene E']
  F [label = 'Gene F']
  G [label = 'Gene G']
  H [label = 'Gene H']
  I [label = 'Gene I']
  J [label = 'Gene J']
  K [label = 'Gene K']
  L [label = 'Gene L']
  M [label = 'Gene M']
  N [label = 'Gene N']
  
  # Set edge defaults
  edge [color = red]

  # Add edges to the graph
  A -> B
  A -> C
  B -> D
  C -> D
  C -> E
  D -> F
  E -> F
  F -> G
  F -> H
  G -> I
  H -> I
  I -> J
  I -> K
  J -> L
  K -> L
  L -> M
  L -> N
}
")


```



## Brick plot
```{r}
suppressWarnings(library(tidyverse))

# Generate mock data
set.seed(123)
metabolites <- paste0("Metabolite", 1:20)
n_endpoints <- sample(1:20, 100, replace = TRUE)
rank <- rank(-n_endpoints)
data <- data.frame(Metabolite = metabolites, N_adverse_events = n_endpoints, Rank = rank)

# Create the brick plot
ggplot(data, aes(x = Rank, y = N_adverse_events, fill = N_adverse_events)) +
  geom_tile() +
  geom_text(aes(label = Metabolite), size = 2, color = "white") +
  scale_fill_gradient(low = "#FDE725", high = "#440154") +
  coord_flip() +
  xlab("Rank") +
  ylab("Number of associated adverse events") +
  theme_bw()


```

**This is a brick plot that shows the ranking of metabolites based on the number of associated adverse events. Each tile represents a metabolite, with the color indicating the number of associated adverse events. The darker the color, the more adverse events a metabolite is associated with. The tiles are ordered by their rank, which is based on the number of adverse events. The x-axis represents the rank, and the y-axis represents the number of associated endpoints. The labels on each tile indicate the name of the corresponding metabolite.**


```{r}
suppressWarnings(library(tidyverse))

# Load the airquality dataset
data(airquality)

# Create a summary table of the number of missing values for each variable
missing_values <- airquality %>%
  summarize_all(~sum(is.na(.))) %>%
  pivot_longer(everything(), names_to = "Variable", values_to = "N_Missing") %>%
  arrange(desc(N_Missing))

# Create the brick plot
ggplot(missing_values, aes(x = rank(-N_Missing), y = N_Missing, fill = N_Missing)) +
  geom_tile() +
  geom_text(aes(label = Variable), size = 3) +
  scale_fill_gradient(low = "yellow", high = "red") +
  coord_flip() +
  xlab("Rank") +
  ylab("Number of missing values") +
  theme_bw()

```


**The x-axis shows the rank of each variable based on the number of missing values, and the y-axis shows the actual number of missing values. Each tile represents a variable, and its color represents the number of missing values it has. The darker the tile, the more missing values the variable has. The plot is useful for identifying which variables have the most missing values and to what extent they affect the data**









## Time series plot
```{r}
suppressWarnings(library(dygraphs))
data(co2)
dygraph(co2, main = "Atmospheric CO2 Concentrations") %>%
  dySeries("V1") %>% #specify which series (column) in the data frame to plot in the dygraph()
  dyRangeSelector()


```




## scatter plot
```{r}
suppressWarnings(library(rbokeh))

# Create sample data
x <- rnorm(100)
y <- rnorm(100)
label <- paste("Point", 1:100 )

# Create plot
p <- figure() %>%
  ly_points(x, y, hover = list(label),color="purple")

# Show plot
p


```


**This is a scatter plot of 100 points, with the x-coordinates and y-coordinates randomly generated from a normal distribution. when hovering over a point, a label indicating the point number is displayed**



```{r}
#install.packages("plotly")
suppressWarnings(library(plotly))

# Select two genes randomly
genes <- sample(rownames(exprs(TCGA_eset)), 2)

# Create a data frame with the gene expression values
plot_data <- data.frame(x = exprs(TCGA_eset)[genes[1], ],
                        y = exprs(TCGA_eset)[genes[2], ])

# Create plot
plot_ly(plot_data, x = ~x, y = ~y, type = "scatter", mode = "markers")

```

**This is a scatter plot of two randomly selected genes from the TCGA_eset dataset. The x-axis represents the expression values of one gene and the y-axis represents the expression values of another gene. Each point on the plot represents a sample from the dataset, where the coordinates of the point are the expression values of the two selected genes for that sample.** 

## MA plot
```{r}
# M values is the difference in expression levels between two groups
logFC <- rnorm(nrow(TCGA_eset), mean = 0, sd = 1)
#A values (mean expression) across two groups
A_value <- rowMeans(exprs(TCGA_eset))
M_value <- logFC

ma_data <- data.frame(Gene = rownames(TCGA_eset),
                      A_value = A_value,
                      M_value = M_value)

ggplot(ma_data, aes(x = A_value, y = M_value)) +
  geom_point(alpha = 0.5) +
  xlab("A - Mean Expression") +
  ylab("M - log Fold Change") +
  theme_bw()

```


**Each point represents a gene, and the position of the point indicates the magnitude of the log fold change and the mean expression. The plot is useful for visualizing differentially expressed genes, as genes with high absolute fold changes are plotted at the top or bottom of the plot, and genes with low mean expression are plotted towards the left or right of the plot. The alpha parameter is used to adjust the transparency of the points in the plot.the x-axis showing the average expression level (A value) and the y-axis showing the log fold change (M value) between two conditions or groups**



## 3D scatterplot
```{r}

suppressWarnings(library(plot3D))
suppressWarnings(library(rgl))
suppressWarnings(library(plotrix))
suppressWarnings(library(Rcmdr))
# Create sample data
x <- rnorm(100)
y <- rnorm(100)
z <- rnorm(100)

# Create 3D scatterplot
scatter3d(x, y, z, main = "3D Scatterplot", xlab = "X", ylab = "Y", zlab = "Z")
```

**The scatterplot shows the distribution of 100 randomly generated values for the X, Y, and Z coordinates in 3D space.** 

